components/eamxx/src/dynamics/homme/eamxx_homme_iop.cpp

```diff
@@ -94,28 +94,40 @@ advance_iop_subsidence(const KT::MemberType& team,
   // Compute updated temperature, horizontal winds, and tracers
   Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_packs), [&] (const int k) {
     auto range_pack = ekat::range<IntPack>(k*Pack::n);
+    const auto at_top = range_pack==0;
+    const auto at_bot = range_pack==nlevs-1;
+    const auto at_mid = not at_top and not at_bot;
 
-    // Get delta(k-1) packs. We need a range pack
-    // that does not contain 0 so that we do not
-    // attempt to access k=-1 index.
-    auto range_pack_m1 = range_pack;
-    range_pack_m1.set(range_pack_m1<1, 1);
+    // Get delta(k-1) packs. The range pack should not
+    // contain index 0 (so that we don't attempt to access
+    // k=-1 index) or index > nlevs-2 (since delta_* views
+    // are size nlevs-1).
+    auto range_pack_for_m1_shift = range_pack;
+    range_pack_for_m1_shift.set(range_pack<1, 1);
+    range_pack_for_m1_shift.set(range_pack>nlevs-2, nlevs-2);
     Pack delta_u_k, delta_u_km1,
-          delta_v_k, delta_v_km1,
-          delta_T_k, delta_T_km1;
-    ekat::index_and_shift<-1>(s_delta_u, range_pack_m1, delta_u_k, delta_u_km1);
-    ekat::index_and_shift<-1>(s_delta_v, range_pack_m1, delta_v_k, delta_v_km1);
-    ekat::index_and_shift<-1>(s_delta_T, range_pack_m1, delta_T_k, delta_T_km1);
-
-    // Get omega_int(k+1) pack. We don't need a specialized
-    // range pack since omega_int contains nlevs+1 entries.
+         delta_v_k, delta_v_km1,
+         delta_T_k, delta_T_km1;
+    ekat::index_and_shift<-1>(s_delta_u, range_pack_for_m1_shift, delta_u_k, delta_u_km1);
+    ekat::index_and_shift<-1>(s_delta_v, range_pack_for_m1_shift, delta_v_k, delta_v_km1);
+    ekat::index_and_shift<-1>(s_delta_T, range_pack_for_m1_shift, delta_T_k, delta_T_km1);
+
+    // At the top and bottom of the model, set the end points for
+    // delta_*_k and delta_*_km1 to be the first and last entries
+    // of delta_*, respectively.
+    if (at_top.any() or at_bot.any()) {
```

Comment:
You may break this into two if statements. There's no point in running the `at_top` set, when only `at_top` has any masked entry, and viceversa.